### DISCUSSION TITLE
Chakra 2.x needs to reduce considerably JS bundle size (with some solutions included)

### DISCUSSION THREAD

**Discussion by TheThirdRace** [chakra-ui/chakra-ui#5277] [Ideas]
## Preface

There's a lot to love in Chakra-UI:

- It's fast & easy to prototype stuff
- Accessibility is taken care of
- The styles are looking good by default
- dark & light theme
- Composition philosophy
- It's flexible enough in most cases
- Documentation is great
- The maintainers and the community are very helpful
- etc.

## Caught between a rock and a hard place

As someone who has been using Chakra almost on a daily basis since version `0.8` (July 2020), I also felt and still feel the pain too...

One of the most frustrating point about Chakra is how a lot of stuff is **NOT** made with the JS bundle size in mind.

There are a couple categories within this:

- Choice of dependencies
- Styling process
- Component design

**The following discussion is about things that could be changed to reduce the JS bundle size considerably.**

*Note: All my bundle size numbers comes from `Webpack Bundle Analyzer` and are expressed in the compressed `gzip` size.*

#### Framer-Motion

The biggest culprit here is `Framer-Motion`, which is a whooping `31.09 KB` when compressed with `gzip`.

To give a bit of perspective, `react-dom` is `37.35 KB` and include the whole React DOM manipulation code with all the bells and whistles.

Another perspective, I can cram the whole `NextJs` framework for `27.54 KB`.

While `Framer-Motion` is a great animation framework, Chakra simply doesn't need it to accomplish what it's supposed to. Let's be real, most of Chakra's animations could each be done with 2 css classes and less than 10 lines of JS. Forcing `31.09 KB` down the users' throat is not only an unsustainable solution, it's wildly inappropriate in this case.

An animation framework is a choice you make when you build an app/website, not for using a UI component library... The choice of using `Framer-Motion` in Chakra is forcing our hand with a very difficult choice:

- Use `Framer-Motion` for our apps/websites and keep our JS bundle "stable", but abnormally high
- Increase even more our bundle by using another animation framework on top and keep a huge amount of dead weight in our JS bundle.

No matter what you choose, this is NOT pretty...

#### Emotion & Styled-System

I'm really confused about what is what here and I would really appreciate an explanation.

To my understanding, `EmotionJs` is the backbone of styling in Chakra, which is fine since it only adds `5.89 KB` for the whole framework.

But then, what is `styled-system` at `8.5 KB`?

If all the styling is done through `EmotionJs`, how come there's a 44% bigger package on the side that looks a lot like styling too?

I'm really confused about that one...

#### Box bundle size

Back in the days, your bundle size would increase by `60 KB` as soon as you add your first Chakra component:

- `<ChakraProvider>` at the top
- the default theme NOT extended (you just import it for future modifications)
- a plain old `<Box>` with "hello world" in it

I can't shake the feeling that `60 KB` for the first `<div>` might be a bit overkill...

I don't know if it has improved since then, but I keep seeing my bundle size increase with pretty much every Chakra release and I'm not using that many components to begin with.

### Styling Process

The styling process is a big pain point for the JS bundle size.

You're encouraged to use style props directly, but this makes your components not reusable between projects and increase the page size considerably because no style is reusable. This is the worst possible scenario for reusability.

If you want to make it a `variant`, then you put it in a NON-treeshakeable variable that contains all the styles for all the Chakra components combined... which increase your bundle size considerably given at this moment there are **57** base components, not counting the ones we create ourselves. This is the worst possible scenario for initial loading performance.

Personally, I've started to remove all the component styles from the Theme to put them in separate `.ts` files that I import when needed. This looks like the best scenario possible, you get reusability out of the box and everything is tree-shakeable which increase loading performance and decrease bundle size for the common & pages parts.

This is something that should be looked at more seriously from Chakra's team.

#### Bundle budget

Is there a bundle budget in Chakra?

Is it managed per component or it's only on `@chakra-ui/react`?

What I do know is that in my common JS bundle, I only use these components:

- ChakraProvider
- EmotionCacheProvider
- Link
- Button
- Menu
- Box
- Icon

Going from version `1.6.8` to `1.7.3` added `7 KB` to my bundle. Remember, this is `7 KB` of compressed JS.

I have no idea if it's all because of 1 component or multiple changes through multiple components, but I doubt it's normal to add `7 KB` just as updates to battle tested components that already worked just fine.

It's really disheartening when you fight tooth and nails to reduce a `3 KB` new feature and then see an update that add 5 KB to 10 KB...

#### Dynamic imports

While I understand not all browsers support `dynamic imports`, every Chakra component should support it from the ground up.

##### Example of good component

For example, `<Drawer>` is awesome because even if it's a heavy component in size, you can easily use a normal `<Button>` to manage the opening and closing.

This allows you to dynamically load `<Drawer>` on a very simple condition, shaving a big `6.5 KB` of JS in your initial bundle.

If the user eventually open the `<Drawer>`, it's downloaded automatically and almost instantly.

If a user don't open the `<Drawer>`, he's never gonna pay the cost.

It's a win-win situation!

##### Example of a very bad component

Now take `<Menu>`... ouch...

You can control the opening by a simple `<Button>`, but that's pretty much it. You can't close it with that same `<Button>`. You can't close it by clicking on the overlay. The focus is not put on the first element either. You can't use arrows to move around in the menu. There's a myriad of stuff not working as intended if you're not using the damn `<MenuButton>`.

The problem with `<MenuButton>` is that it's located within `<Menu>` itself:

```typescript
<Menu>
  <MenuButton />
  <MenuList />
</Menu>
```

That little detail makes it almost impossible to use `dynamic import`. By itself, `<Menu>` is around `13.7 KB`. Again don't forget it's all `gzip` sizes, which is HUGE!

There's so much stuff in `<Menu>`, including `PopperJs` and multiple providers, that by the time you get to that button, the `dynamic import` would save you around `0.3 KB`...

And your users pay that cost on first load no matter what, even if they never open that menu.

I was actually able to use `dynamic import` and shave that `14 KB` for the whole menu. Here's the horror show:

1. Create a component with the whole `<Menu>` in it
2. In parent component
   1. Use `dynamic import()` to import the `<Menu>` component you just created
   2. Add a state to track which step your are in the process of loading dynamically your `<Menu>`
   3. Add a dummy `<Button>` with an `onClick` that set the tracking state to `1`
   4. A `useEffect` triggers on `1` to load dynamically the `<Menu>` and once loaded, it will set tracking state to `2`
   5. Once tracking state is `2` the `<Menu>` will render with a `ref` that will be passed to `<MenuButton>` and the dummy `<Button>` will be removed
   6. A `useEffect` triggers on `2` to focus on the new `<MenuButton>` and **click on it**

This is a terribad workaround... We can all agree on that. But this allows me to switch the dummy `<Button>` on the fly for that freaking `<MenuButton>` as soon as I click on it. It doesn't have any flash of element disappearing then reappearing, no CLS (Cumulative Layout Shift) and it works perfectly. It ain't pretty, but that ain't my fault either...

I should be able to fully control `<Menu>` from outside of it. But given a lot of features a tied down to `<MenuButton>` and it's located inside `<Menu>`, you're screwed. This is a bad design if you care about loading performance and bundle size.

## Conclusion

So what should we get from this long text?

I think one of the major goal of Chakra `2.x` should be to reduce its bundle size.

It would allow:

- Better flexibility, design and composition
- Better performance
- Better overall styling size and especially better typings for the styles

If I had time, I would clearly contribute to Chakra by making PRs. Unfortunately, I don't have enough time for that kind of involvement. So all I can do is start a discussion about it and take some time to answer, or debate, the questions that could arise from such discussions.

**Comment by hbendix** [comment]
By simply adding `ChakraProvider` to `_app.tsx` as suggested [here](https://chakra-ui.com/guides/getting-started/nextjs-guide#provider-setup) adds a whopping 45kb for `chakra-ui` and an additional 32kb for `framer-motion`, how is this acceptable for production? 

![Screenshot 2022-04-21 at 14 15 22](https://user-images.githubusercontent.com/24959221/164455906-14312be2-a719-4b8c-af44-43ca0603fe28.png)

**Reply by TheThirdRace** [reply]
And that's before you include your own theme values for your components.

Anything that is a `variant` is loaded in the `theme` object in the `ChakraProvider` for all components that exists even if you don't use them in the page. And it's not out of the realm of possibilities to have around 10 to 25 KB of CSS... 

The cost is extreme for production bundles, I'm seriously considering dropping Chakra entirely because it's simply unsustainable in the long run.

**Reply by runoncedev** [reply]
I get the main point but as someone whos coming from MUI, Chakra UI feels really lightweight in comparison.

**Reply by b2m9** [reply]
> I get the main point but as someone whos coming from MUI, Chakra UI feels really lightweight in comparison.

Sure, that might very well be. But "the other option is way worse" is not a good argument for anything.

**Comment by bhaskar20** [comment]
this is a major pain point for me, team/company is thinking of avoiding the chakra-ui from future projects altogether

**Comment by isaachinman** [comment]
I'd like to add that Chakra's reliance on `emotion` means that you _cannot_ render via SSG with NextJs. This was indeed a very poor choice, and I was disappointed to discover it.

**Reply by TheThirdRace** [reply]
@isaachinman

Could you elaborate please?

I've been using SSG with NextJs and Chakra for 2 years now, but I've yet to encounter any problem. Maybe I'm missing something?

**Reply by SheaBelsky** [reply]
Chakra is designed to be compatible with Next.js' server-side rendering capabilities. Check out this article: https://chakra-ui.com/guides/getting-started/nextjs-guide

**Reply by isaachinman** [reply]
@SheaBelsky I've seen the doc. Can you point to where in the codebase the critical emotion styles are generated for SSG/`getStaticProps`?

**Reply by SheaBelsky** [reply]
I might have misunderstood your initial question - Is your concern that emotion styles are always calculated on the client, rather than being calculated on the server once and then passed down to the client? Or is your concern that SSG outright does not work with Chakra + Next.js?

**Reply by isaachinman** [reply]
> Is your concern that emotion styles are always calculated on the client, rather than being calculated on the server once and then passed down to the client

Yes, correct. If critical styles aren't calculated during SSG, FOUC will occur. Emotion's current NextJs guides all entail use of `getInitialProps` on a custom `_document`, which is _not_ SSG.

**Reply by TheThirdRace** [reply]
@isaachinman 
There's something odd about what your last comment:

> Emotion's current NextJs guides...

There might be a slight nuance you should make. We're talking here of the use of Chakra UI, which use EmotionJs under the hood, within NextJs SSG.

I just looked at the HTML document downloaded on the first network request for my website and all the styles from SSG are included in it, already calculated correctly.

At the moment, it's true that you see FOUC (Flash Of Unstyled Content) with Chakra UI, but that's not because it's impossible to fix. I think it's related to either ChakraUI not always using their own tokens (which are all linked to CSS Custom Properties, meaning it should work as-is) or do some weird theming stuff in their theme provider. My last debug session on the matter pointed toward their Theme Provider going back and forth between light/dark theme... I see no reason for the whole thing to work flawlessly without FOUC if Chakra UI sticks to their theme tokens and stops messing with the color mode (light/dark) in their Provider.

I might be wrong, but it just looks like a bug on their part, not something that is impossible to do.

**Reply by ivan-kleshnin** [reply]
I'm using Chakra with NextJS' SSG and I have no FOUC.

> Emotion's current NextJs guides all entail use of getInitialProps on a custom _document, which is not SSG.

Here is your mistake. `getInitialProps` on a `_document` does not prevent SSG (it's mentioned in the docs, I can't quickly recall the exact page), `_document` is very a special component with its own rules.

**Comment by whatsong** [comment]
Huge pain to be honest. The bundle size keeps increasing with each version, and there is nothing you can do about it due to poor design. Such a shame because I love the Chakra UI, it's honestly the best out there. However I'm at the point now where I am going to have to remove it from all my production apps, as no thought has been put into how to solve this problem and it just keeps getting worse. 

And yes, why on earth is it using Framer!! I don't want to be forced into such a large dependency, especially when it's used in 0.1% of chakra components. Could so easily be done with css. 

Does anyone know if anything is being done to rectify any of this? I'm sort of stuck between a rock and a hard place, as replacing the UI with another system is obviously a huge job. However my first-page loads are over 300kb and growing due to Chakra, which is insane, especially as so much emphasis is put on page speed and first load times, by Google over the last few years.

In each of my apps that use Chakra -> SEO, speed and usability are suffering. Not only suffering, but getting worse with each version, instead of better.

**Reply by TheThirdRace** [reply]
I'm not privy to what Chakra is planning, but they've just released [ZagJs](https://zagjs.com/).

This might not seem related, but I think it's closely linked to the bundle size problem.

Chakra is recognizing they needed to extract their components logic into machine state. Not only will it help them port their components to other frameworks more easily, but it should greatly simplify the internal logic for some components. I'm extrapolating, but the results should be less bugs and a smaller JS footprint.

I just hope that while extracting their components logic, they will take the opportunity to rewrite some components to be more easily tree-shakable.

As for their use of Framer Motion, it's been brought to the spotlight multiple times. I've had a chat with a member of the team a while ago and they were reluctant to remove it because it would be a breaking change and they didn't want to release a `2.x` version yet...

But I don't see this reasoning holding anymore since React 18 dropped not long ago and they had to "urgently" release a `2.x`. Given they're extracting their component logic at the moment, it would be a golden opportunity to boot Framer Motion out...

It would be nice if Chakra would be more outspoken about their plans. I saw some roadmap somewhere, but can't find it anymore ☹️

**Reply by ivan-kleshnin** [reply]
It's curios that Zag examples include SolidJS. This little cool thing is getting traction. 🤔

**Comment by taitems** [comment]
This may or may not be in the spirit of open source, but have you considered a bounty? 

There are two potential models:
- Enterprises and passionate individuals can "fund" a high priority issue to get it addressed in a timely fashion
- The Chakra UI team itself could put a bounty up for contributors to claim if they can fix it themselves

Ultimately both make me a bit uncomfortable but this issue may not be getting the attention it deserves without it?

**Comment by n1kio** [comment]
I certainly would put some money into the bounty. I think with the current package size chakra can't be taken serious in some contexts.

**Comment by veloware** [comment]
Similar discussions in https://github.com/chakra-ui/chakra-ui/issues/4975, and there seems to be some related pull requests being created / merged for refactoring the theme to reduce the bundle size.

**Comment by lhguerra** [comment]
I managed to get rid of a good amount of js by creating my custom chakra provider. The default one is this: https://github.com/chakra-ui/chakra-ui/blob/main/packages/components/react/src/chakra-provider.tsx

I just removed the toast provider and saved 36kb in my build chunk!
Since I'm not using toast anywhere this is not an issue rn, and since this is such a heavy feature I think it should be a separate package or something like that.

**Reply by emirsavran** [reply]
I guess it's about `framer-motion`. You probably do not import anything that depends on `framer-motion` except the toast.
https://github.com/chakra-ui/chakra-ui/search?q=framer-motion&type=

**Reply by MigHerCas** [reply]
@lhguerra Could you share your approach? I tried to do the same but somehow `useColorMode` hook isn't working anymore. Thanks beforehand

+ I could save ~`30kb` in my build, so this is a good idea

Edit: I had to upgrade `chakra-ui/react` to `^2.4.9`, it's working now

**Comment by levi0214** [comment]
This should be treated as 1st priority, otherwise this lib would gradually die.
I'm already considering to switch to some custom ui code.

**Reply by thefrana** [reply]
I think we should all appreciate the work of @segunadebayo and other contributors. Yes, the size is a problem, but the Chakra team is working hard on delivering four new "legs" of the project like state machines, framework interoperability or better CSS solution. All this should decrease bundle sizes. Just be patient and thank them for what they already done.

That being said, with some minor tweaking, you can get score 100 with Chakra on Lighthouse. The framework has beautiful DX that superseded everything I used in the past.

**Reply by pablocubico** [reply]
Hey @lksfrn (or anyone!) any ideas on those tweaks? I'm using NextJS, and I tried everything, I think I'll do a custom provider to remove the toast as someone mentioned here. Any other suggestions to get a good Lighthouse score? It's killing me on the mobile report.

**Reply by thefrana** [reply]
@pablocubico NextJS with ChakraUI should give you by default green score on mobile devices. Just by using any UI library, you are opting to their trade-offs in exchange for shorter development time. Alternatives to ChakraUI are much worse in this regard. Lighthouse score is combinations of different factors. I never had problems with performance score, so I cannot help you in this regard.

**Reply by bmlyon** [reply]
If we're trying to reduce overall application bundle size wherever we can, and Chakra has a suboptimal bundle size, it's not helpful when someone says "Chakra (on it's own) has a green Lighthouse score". Chakra's bundle size should be better optimized.

**Reply by levi0214** [reply]
@lksfrn You are right. Sorry for my complaints. That won't help. Maybe it's better to make a contribution instead.

**Reply by lhguerra** [reply]
@xlukasfrana the thing is CWV have kicked in almost two years ago, and we still struggle with the huge amount of js that comes together just by adding the chakra provider to our apps. I recently gave up and switched to theme-ui, which I think is much worse than chakra for development experience but still better than getting yelled at every week by my clients.

**Comment by proustibat** [comment]
This the only reason why I can't use it in a professional environment, even if I like it

**Comment by n1kio** [comment]
No offense and I’m very thankful for all developers, but I can’t understand how this issue is not priority number one for Chakra. :/I‘m certain a lot companies would be happy to fund this if resources are the issue. I would contribute. Am 22.03.2023 um 15:32 schrieb Luiz Henrique Guerra ***@***.***>:﻿
@xlukasfrana the thing is CWV have kicked in almost two years ago, I we still struggle with the huge amount of js that comes together just by adding the chakra provider to our apps. I recently gave up and switched to theme-ui, which I think is much worse than chakra for development experience but still better than getting yelled at every week by my clients.

**Comment by zachguo** [comment]
Before we get v3, is there a v2 fork that removes Toast, framer-motion and all the social themes?

**Comment by isaachinman** [comment]
Would again like to float the idea of a bounty – I can kick in a portion of it. I'd also be happy to help with dev work in any way.

**Comment by rmannibucau** [comment]
+1, trying to reduce my bundle but seems some are hard without rebundling @chakra-ui/react (like framer). Sounds like a quick win for a new 2.x.

**Comment by TheThirdRace** [comment]
Just saw this article about Chakra's future: https://www.adebayosegun.com/blog/the-future-of-chakra-ui

It looks very good for bundle size. I think the Chakra team took seriously the problem we reported. It might have taken a while to formulate a plan and enact it, lots of pieces were needed, but I think it bodes well for Chakra's future.
